### ISSUE TITLE
OPENVLA finetuned on custom dataset have very low performance

### ISSUE THREAD

**Issue by reinaldoyang** [openvla/openvla#304]
Hi! thank you for taking the time to read this issue.
I'm currently creating my own dataset in pybullet environment collected at 5hz, and converted this dataset to RLDS format for further fine tuning.
I'm currently randomize the position of the cube and the tray on the working area(marked by the red rectangular area)
I collected around 210 demonstrations to grasp the cube and put it in the tray

<img width="1008" height="615" alt="Image" src="https://github.com/user-attachments/assets/eb04f8e0-d51b-4b3a-bf47-93fc73b6a33c" />
Below is my training parameters:
GPU: 5090 32gb

``` bash
torchrun --standalone --nnodes 1 --nproc-per-node 1 vla-scripts/finetune.py 
--vla_path "openvla/openvla-7b" 
--data_root_dir ~/tensorflow_datasets 
--dataset_name robot_dataset 
--run_root_dir ~/openvla_runs/robot_experiment 
--adapter_tmp_dir ~/openvla_runs/robot_experiment/adapters 
--lora_rank 32 
--batch_size 4 
--grad_accumulation_steps 8 
--learning_rate 5e-4 
--image_aug False 
--wandb_project 
--wandb_entity
--save_steps 500
--max_steps 10000
``` 
and for the wandb logging, it says that the model already converges early

<img width="1722" height="488" alt="Image" src="https://github.com/user-attachments/assets/b705e59a-2175-4647-a454-ee2952d33e1e" />
but for the inference, the robot still can't seem to grasp and move the cube to the tray, some occasions it can grasp the cube, and sometimes it can grab and put the cube inside the tray (very low success rate)

my questions is, is it because of my dataset is just not large enough? or maybe my evaluation script is counted as zero-shot inference? because i created a new randomized position of cube and tray? or should i set my evaluation script to check within the training data distribution?

**Comment by lovecode1** [comment]
First of all, the loss graph only shows that the model is predicting the correct action on the training data with a low error rate. It does not guarantee that the fine-tuning will generalize well to new data.

Try saving screenshots when you run the model on real data, then log the actions and check whether the predicted actions make sense.

Why did you set image_aug to false?

I will also change the camera position and orientation between episodes to create a greater variety of data.

How many steps do you have in each episode?

You could also try replacing the cube with a different object, as well as swapping out the tray. This can help the model generalize better.

Additionally 10k steps might not be enough.

**Comment by reinaldoyang** [comment]
1. For the evaluation, I set the environment to GUI mode, so I can see what's going on. The way the robot approach the cube makes sense, but it sometimes just doesn't grasp successfully.
2. I'm not sure whether image_aug should be set to false or true
3. Because my evaluation script also camera position and orientation are the same as the training data, so I didn't bother to change it
4. I have around 20 steps in each episodes
5. For my evaluation script, the objects are also a tray and a cube, so I did not replace the cube with other objects, so I'm not sure if that will really help with the cube grasping
6. I'll try more training steps

**Comment by lovecode1** [comment]
The main goal in training or fine-tuning an AI model is to ensure it doesn’t memorize but rather generalizes. This way, even when it encounters a new angle or position, it can still determine the correct next move.

By switching camera positions and angles, enabling image augmentation, and adjusting other basic aspects of the scene, you force the model to learn the tasks regardless of its surroundings.

When all the training data is too similar, the model tends to memorize those details. As a result, when the environment changes even slightly, it may fail to respond correctly.

For example, if the model faces a move or an angle it has never seen before, and it is not generalized enough, it will not know what to do next.

**Comment by lovecode1** [comment]
You also need to ensure that the action data matches the robot’s coordinate system etc. and that there are no mismatches. Try running the training data actions directly on the robot to verify that it moves in the way you expect. This should be done even before any training takes place. It’s important to first confirm that the training data itself is correct.

**Comment by reinaldoyang** [comment]
I've ensured that the data collection pipeline is correct, I've go through the collected data already. I'll try to add more steps for training.
You mentioned that if the model faces an unseen move or angle it will fail, so doesn't that mean there's no need to vary camera placement? because the current evaluation setup is exactly the same as training, rather maybe increase the variance of robot states?

**Comment by lovecode1** [comment]
I think you’re missing the point of an AI model.
The goal is not to simply record a sequence of steps and replay them. Instead, you are training the AI to understand the concept of picking up an item and placing it in a different location.

Imagine if, for your entire life, you could only see the world from one fixed angle and position, with no ability to explore. Your knowledge would be extremely limited, and there would be many tasks you couldn’t complete.

In the same way, an AI model needs variety. Without variance, it will only memorize specific steps rather than learning the underlying concept.

**Comment by reinaldoyang** [comment]
Will close this issue, I found out the problem was because there are 2 consecutive frames of image observation with the same robot states. Tips for others encountering the same issue is to make your problem as simple as possible so it's easier to debug. I first reduced the spawn area of those cube and tray to make the spatial problem simpler, and I didn't change any camera angle, etc (reduce any possible variable)

**Comment by vanstrong12138** [comment]
> Hi! thank you for taking the time to read this issue.嗨！感谢您花时间阅读本期内容。 I'm currently creating my own dataset in pybullet environment collected at 5hz, and converted this dataset to RLDS format for further fine tuning.我目前正在 pybullet 环境中创建自己的数据集，收集 5Hz 频率，并将该数据集转换为 RLDS 格式以便进一步微调。 I'm currently randomize the position of the cube and the tray on the working area(marked by the red rectangular area)我目前正在随机调整方块和托盘在工作区（用红色矩形区域标记）的位置 I collected around 210 demonstrations to grasp the cube and put it in the tray我收集了大约210个演示，准备把立方体放进托盘
> <img alt="Image" width="822.3333740234375" height="430.16668701171875" src="https://private-user-images.githubusercontent.com/104484204/490872488-eb04f8e0-d51b-4b3a-bf47-93fc73b6a33c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzA3OTMxNDMsIm5iZiI6MTc3MDc5Mjg0MywicGF0aCI6Ii8xMDQ0ODQyMDQvNDkwODcyNDg4LWViMDRmOGUwLWQ1MWItNGIzYS1iZjQ3LTkzZmM3M2I2YTMzYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMjExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDIxMVQwNjU0MDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04MTI3MTZkYjQ4NzRlMTc3MTcwMzdiMWQ1ZTE5NzFjN2M2MDRmZWQzYzIzYjYxMWExMGYwODNlMjAwMjQ1M2I4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.vt1NvE--H7XkKWQrozGMFUPKWjeFPrGVFXIqcVAFPQ8"> Below is my training parameters: GPU: 5090 32gb
> [ <img alt="Image" width="822.3333740234375" height="430.16668701171875" src="https://private-user-images.githubusercontent.com/104484204/490872488-eb04f8e0-d51b-4b3a-bf47-93fc73b6a33c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzA3OTMxNDMsIm5iZiI6MTc3MDc5Mjg0MywicGF0aCI6Ii8xMDQ0ODQyMDQvNDkwODcyNDg4LWViMDRmOGUwLWQ1MWItNGIzYS1iZjQ3LTkzZmM3M2I2YTMzYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMjExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDIxMVQwNjU0MDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04MTI3MTZkYjQ4NzRlMTc3MTcwMzdiMWQ1ZTE5NzFjN2M2MDRmZWQzYzIzYjYxMWExMGYwODNlMjAwMjQ1M2I4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.vt1NvE--H7XkKWQrozGMFUPKWjeFPrGVFXIqcVAFPQ8"> ](https://private-user-images.githubusercontent.com/104484204/490872488-eb04f8e0-d51b-4b3a-bf47-93fc73b6a33c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzA3OTMxNDMsIm5iZiI6MTc3MDc5Mjg0MywicGF0aCI6Ii8xMDQ0ODQyMDQvNDkwODcyNDg4LWViMDRmOGUwLWQ1MWItNGIzYS1iZjQ3LTkzZmM3M2I2YTMzYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMjExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDIxMVQwNjU0MDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04MTI3MTZkYjQ4NzRlMTc3MTcwMzdiMWQ1ZTE5NzFjN2M2MDRmZWQzYzIzYjYxMWExMGYwODNlMjAwMjQ1M2I4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.vt1NvE--H7XkKWQrozGMFUPKWjeFPrGVFXIqcVAFPQ8) 以下是我的训练参数： 显卡：5090 32GB
> 
> torchrun --standalone --nnodes 1 --nproc-per-node 1 vla-scripts/finetune.py 
> --vla_path "openvla/openvla-7b" 
> --data_root_dir ~/tensorflow_datasets 
> --dataset_name robot_dataset 
> --run_root_dir ~/openvla_runs/robot_experiment 
> --adapter_tmp_dir ~/openvla_runs/robot_experiment/adapters 
> --lora_rank 32 
> --batch_size 4 
> --grad_accumulation_steps 8 
> --learning_rate 5e-4 
> --image_aug False 
> --wandb_project 
> --wandb_entity
> --save_steps 500
> --max_steps 10000
> 
> and for the wandb logging, it says that the model already converges early对于 WANDB 日志，它说模型已经很早收敛了
> <img alt="Image" width="822.3333740234375" height="233.0500030517578" src="https://private-user-images.githubusercontent.com/104484204/490888213-b705e59a-2175-4647-a454-ee2952d33e1e.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzA3OTMxNDMsIm5iZiI6MTc3MDc5Mjg0MywicGF0aCI6Ii8xMDQ0ODQyMDQvNDkwODg4MjEzLWI3MDVlNTlhLTIxNzUtNDY0Ny1hNDU0LWVlMjk1MmQzM2UxZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMjExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDIxMVQwNjU0MDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ZWFjYTQwMTk4ZjRhOTEzMjQ1YzIwYTIzOTM3M2E2ZTNkZWEzYWU2NDJmYTlmMzhkMzExNWIwMGM2NGYzMWExJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.s2e4pKSD4PFscZAJmfL3UEXQqCfJBM6Dj3fL-_BUkR4"> but for the inference, the robot still can't seem to grasp and move the cube to the tray, some occasions it can grasp the cube, and sometimes it can grab and put the cube inside the tray (very low success rate)
> [ <img alt="Image" width="822.3333740234375" height="233.0500030517578" src="https://private-user-images.githubusercontent.com/104484204/490888213-b705e59a-2175-4647-a454-ee2952d33e1e.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzA3OTMxNDMsIm5iZiI6MTc3MDc5Mjg0MywicGF0aCI6Ii8xMDQ0ODQyMDQvNDkwODg4MjEzLWI3MDVlNTlhLTIxNzUtNDY0Ny1hNDU0LWVlMjk1MmQzM2UxZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMjExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDIxMVQwNjU0MDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ZWFjYTQwMTk4ZjRhOTEzMjQ1YzIwYTIzOTM3M2E2ZTNkZWEzYWU2NDJmYTlmMzhkMzExNWIwMGM2NGYzMWExJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.s2e4pKSD4PFscZAJmfL3UEXQqCfJBM6Dj3fL-_BUkR4"> ](https://private-user-images.githubusercontent.com/104484204/490888213-b705e59a-2175-4647-a454-ee2952d33e1e.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzA3OTMxNDMsIm5iZiI6MTc3MDc5Mjg0MywicGF0aCI6Ii8xMDQ0ODQyMDQvNDkwODg4MjEzLWI3MDVlNTlhLTIxNzUtNDY0Ny1hNDU0LWVlMjk1MmQzM2UxZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMjExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDIxMVQwNjU0MDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ZWFjYTQwMTk4ZjRhOTEzMjQ1YzIwYTIzOTM3M2E2ZTNkZWEzYWU2NDJmYTlmMzhkMzExNWIwMGM2NGYzMWExJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.s2e4pKSD4PFscZAJmfL3UEXQqCfJBM6Dj3fL-_BUkR4) 但从推断角度来看，机器人仍然无法抓住并移动立方体到托盘上，有时能抓住立方体，有时能抓取并放入托盘（成功率非常低）
> 
> my questions is, is it because of my dataset is just not large enough? or maybe my evaluation script is counted as zero-shot inference? because i created a new randomized position of cube and tray? or should i set my evaluation script to check within the training data distribution?我的问题是，是不是因为我的数据集不够大？或者我的评估脚本被算作零样本推断？因为我创建了一个新的随机立方体和托盘位置？还是我应该让评估脚本在训练数据分布中检查？

Could you tell me how to build my own dataset? I'm a complete beginner when it comes to VLA.
